common/exclusiveStringMap.go

```diff
@@ -0,0 +1,82 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"github.com/pkg/errors"
+	"strings"
+	"sync"
+)
+
+type ExclusiveStringMap struct {
+	lock          *sync.Mutex
+	m             map[string]struct{}
+	caseSensitive bool
+}
+
+func NewExclusiveStringMap(caseSensitive bool) *ExclusiveStringMap {
+	return &ExclusiveStringMap{
+		lock:          &sync.Mutex{},
+		m:             make(map[string]struct{}),
+		caseSensitive: caseSensitive,
+	}
+}
+
+var exclusiveStringMapCollisionError = errors.New("cannot simultaneously send two files to same destination name")
+
+// Add succeeds if and only if key is not currently in the map
+func (e *ExclusiveStringMap) Add(key string) error {
+	key = e.convertCase(key)
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	_, alreadyThere := e.m[key]
+	if alreadyThere {
+		return exclusiveStringMapCollisionError
+	}
+	e.m[key] = struct{}{}
+	return nil
+}
+
+func (e *ExclusiveStringMap) Remove(key string) {
+	key = e.convertCase(key)
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	delete(e.m, key)
+}
+
+func (e *ExclusiveStringMap) convertCase(s string) string {
+	if e.caseSensitive {
+		return s
+	} else {
+		// technically, toLower is a bad way to make something case insensitive,
+		// but, in our case, the collisions we are concerned about are
+		// ones that are caused due to US (our code) mutating characters that fall within the ASCII
+		// section of the code space (specifically escaping or unescaping special characters not
+		// supported by Windows or Azure Files); rather than general case sensitivity issues in the broader sense.
+		// So we think this is OK.
+		// TODO: double check, is it really? Should we replace by going through rune-by-rune converting to unicode "fold case"?
```

Comment:
> TODO: double check, is it really? [](start = 4, length = 34)

hmmm, I also feel like it's ok 

[https://golang.org/pkg/strings/#ToLower](https://golang.org/pkg/strings/#ToLower) says each unicode letter is mapped to their lower case, so I guess we are fine

Reply:
It's not quite that simple, unfortunately, in some cases.   E.g. where there are muliple ways to represent one thing.  Maybe we should change it to work through rune-by-rune, but I didn't want to over complicate things....

Reply:
Technically there are cases where the number of characters in the lower and uppercase form of a "thing" are not the same.  E.g. German ß and SS are considered equal in some case insensitive comparisons…. But, not by Windows. Even with German set as the OS language, it treats ß.txt and SS.txt as separate files in a test I ran just now.

It also let me have files named İ.txt and i.txt even though, in Turkish, those differ only by case. (I didn't actually try setting Turkish as the Windows language, for fear of getting lost)

So, on reflection, I think I basically agree with you.  I feel that what we are doing with lowercasing is probably as close as we can realistically get to what the OS will do. (And that's what we need to match, not what native speakers of a language would do. This is all about how we interact with the OS - so if OS and humans disagree we should align, in this case, to the OS).

Reply:
Does Go convert ß to SS in this scenario? If so, that's kind of worrisome. We have two places this will cover-- Unsafe characters on Windows/Files, and decompression. There's obviously no way we're going to cover EVERY edge case in this context, if Go doesn't follow the OS's interpretation of upper/lowercase characters.

---
In reply to: [331329480](https://github.com/Azure/azure-storage-azcopy/pull/638#discussion_r331329480) [](ancestors = 331329480)

Reply:
Does Go follow the OS's interpretation of upper/lower case, or does it use its' own specific implementation?

If it uses its' own specific implementation, I can forsee a scenario in which the implementations collide.

There's an OK situation: The OS doesn't consider two characters to be upper/lowercase variants of eachother, but Go does. In this case, a file just errors out. No biggie.

Then there's a worrisome situation: Go doesn't consider two characters to be upper/lowercase variants of eachother, but the OS does. Then, we get our old corruption issue again.

Reply:
I agree with your logic.  A possible solution is to write our own routine to go through each string rune by rune and turn it into a folded, comparable case.  (I think to be strictly correct you must do unicode normalization before that). 

But, I can't get that done for the release date.  I'll make a backlog item to evaluate in more detail once we have time.